mypy/subtypes.py

```diff
@@ -63,6 +63,33 @@ def is_subtype(left: Type, right: Type,
     between the type arguments (e.g., A and B), taking the variance of the
     type var into account.
     """
+    if (isinstance(left, TypeAliasType) and isinstance(right, TypeAliasType) and
+            left.is_recursive and right.is_recursive):
+        # This case requires special care because it may cause infinite recursion.
```

Comment:
Probably worth documenting this in more detail.

To make sure I understand this, the algorithm is basically "under the assumption that left <: right, check that get_proper_type(left) <: get_proper_type(right)"?

(Might be worth splitting out some of the logic into another function again, but your call)

Reply:
> To make sure I understand this, the algorithm is basically "under the assumption that left <: right, check that get_proper_type(left) <: get_proper_type(right)"?

Yes. I will add some docs for this.